src/Orleans.ScheduledJobs/LocalScheduledJobManager.cs

```diff
@@ -0,0 +1,362 @@
+                this,
+                CancellationToken.None,
+                TaskCreationOptions.None,
+                WorkItemGroup.TaskScheduler).Unwrap();
+            _listenForClusterChangesTask.Ignore();
+        }
+
+        LogStarted(_logger);
+        return Task.CompletedTask;
+    }
+
+    private async Task Stop(CancellationToken ct)
+    {
+        LogStopping(_logger, _runningShards.Count);
+        
+        _cts.Cancel();
+        
+        if (_listenForClusterChangesTask is not null)
+        {
+            await _listenForClusterChangesTask;
+            //_listenForClusterChangesTask = null;
+        }
+        
+        await Task.WhenAll(_runningShards.Values.ToArray());
+        
+        // Dispose any remaining shards in the cache
+        foreach (var shard in _shardCache.Values)
+        {
+            try
+            {
+                await shard.DisposeAsync();
+            }
+            catch (Exception ex)
+            {
+                LogErrorDisposingShard(_logger, ex, shard.Id);
+            }
+        }
+        
+        LogStopped(_logger);
+    }
+
+    private async Task ProcessMembershipUpdates()
+    {
+        await Task.CompletedTask.ConfigureAwait(ConfigureAwaitOptions.ForceYielding | ConfigureAwaitOptions.ContinueOnCapturedContext);
+        var current = new HashSet<SiloAddress>();
+        await foreach (var membershipSnapshot in _clusterMembershipUpdates.WithCancellation(_cts.Token))
+        {
+            try
+            {
+                // Get active members.
+                var update = new HashSet<SiloAddress>(membershipSnapshot.Members.Values
+                    .Where(member => member.Status == SiloStatus.Active)
+                    .Select(member => member.SiloAddress));
+
+                // If active list has changed, track new list and notify.
+                if (!current.SetEquals(update))
+                {
+                    current = update;
+                    await GetUnassignedShards();
+                }
+            }
+            catch (Exception exception)
+            {
+                LogErrorProcessingClusterMembership(_logger, exception);
+            }
+        }
+    }
+
+    private async Task GetUnassignedShards()
+    {
+        LogCheckingForUnassignedShards(_logger);
+        
+        var shards = await _shardManager.AssignJobShardsAsync(this.Silo, DateTime.UtcNow.AddHours(1), _cts.Token);
+        if (shards.Count > 0)
+        {
+            LogAssignedShards(_logger, shards.Count);
+            foreach (var shard in shards)
+            {
+                StartRunningShardTracked(shard);
+            }
+        }
+        else
+        {
+            LogNoShardsToAssign(_logger);
+        }
+    }
+
+    private void StartRunningShardTracked(IJobShard shard)
+    {
+        LogStartingShard(_logger, shard.Id, shard.StartTime, shard.EndTime);
+        _runningShards.TryAdd(shard.Id, RunShard(shard));
```

Comment:
We should use logic to check that we are not already running the shard here, so that this can be idempotent